plutus-ledger-api/src/v1/value.rs

```diff
@@ -186,11 +355,79 @@ impl Value {
                 .collect(),
         )
     }
+
+    /// Create a vector with each distinct value
+    pub fn flatten(&self) -> Vec<(&CurrencySymbol, &TokenName, &BigInt)> {
+        self.0
+            .iter()
+            .flat_map(|(currency_symbol, assets)| {
+                assets
+                    .iter()
+                    .map(move |(token_name, amount)| (currency_symbol, token_name, amount))
+            })
+            .collect()
+    }
+
+    pub fn unflatten(list: &[(CurrencySymbol, TokenName, BigInt)]) -> Self {
+        list.iter()
+            .fold(Value::new(), |v, (cs, tn, am)| v.insert_token(cs, tn, am))
+    }
+}
+
+impl fmt::Display for Value {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut it = self
+            .0
+            .iter()
+            .flat_map(|(currency_symbol, assets)| {
+                assets
+                    .iter()
+                    .map(move |(token_name, amount)| (currency_symbol, token_name, amount))
+            })
+            .peekable();
+        while let Some((cur_sym, tn, amount)) = it.next() {
+            if tn.is_empty() {
+                write!(f, "{} {}", cur_sym, amount)?;
+            } else {
+                write!(f, "{}.{} {}", cur_sym, tn, amount)?;
+            }
+            if it.peek().is_some() {
+                write!(f, "+")?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub(crate) fn flat_value(
+    input: &str,
+) -> IResult<&str, (CurrencySymbol, TokenName, BigInt), VerboseError<&str>> {
+    let (input, cs) = currency_symbol(input)?;
+
+    let (input, tn) = alt((preceded(char('.'), token_name), success(TokenName::ada())))(input)?;
+
+    let (input, amount) = preceded(char(' '), big_int)(input)?;
+
+    Ok((input, (cs, tn, amount)))
+}
+
+pub(crate) fn value(input: &str) -> IResult<&str, Value, VerboseError<&str>> {
```

Comment:
Leave some doc strings for the parsers please: the formats of strings they expect? some examples? It doesn't need to be formal but should help people understand what's happening at a glance.